### PR TITLE
add cue sheet support

### PR DESCRIPTION
**Pull request by lifegpc**
#687 
[test file](https://1drv.ms/u/s!Aoq8mAMlVv8UnSQnNf9ZhhGWLGBB?e=ELBC3q) generated by EAC.
Now it support all 4 types. 
BUT IT DON'T SUPPORT GAP.

### CODE REVIEW

File: src/mpc-hc/PlayerPlaylistBar.cpp [hunk 1]
```diff
@@ -402,6 +405,150 @@ bool CPlayerPlaylistBar::ParseBDMVPlayList(CString fn)
...
+    while(f.ReadString(str)) {
+        str.Trim();
+        if (cue_index == 0 && str.Left(5) == _T("TITLE")) {
+            title = str.Mid(6).Trim(_T("\""));
+        }
+        else if (cue_index == 0 && str.Left(9) == _T("PERFORMER")) {
+            performer = str.Mid(10).Trim(_T("\""));
+        }
+        else if (str.Left(4) == _T("FILE")) {
+            if (str.Right(4) == _T("WAVE") || str.Right(4) == _T("MP3") || str.Right(4) == _T("AIFF")) { // We just support audio file.
```
Comment (clsid2) [inline src/mpc-hc/PlayerPlaylistBar.cpp]: str.Right(4) == _T("MP3")

Should be 3

Comment (clsid2) [reply src/mpc-hc/PlayerPlaylistBar.cpp]: Are more formats possible? Like FLAC?

**Comment by clsid2** [comment]
I will check it more in depth later. Things can probably be simplified a bit.

Maybe done already. If a file has just one chapter, then pli.m_cue (and related values) should be set to false. To avoid unnecessary parsing of cue again later.

File: src/mpc-hc/PlayerPlaylistBar.cpp (hunk 1 above)
Comment (lifegpc) [reply src/mpc-hc/PlayerPlaylistBar.cpp]: It should be ```WAVE```.

Comment (lifegpc) [reply src/mpc-hc/PlayerPlaylistBar.cpp]: ```flac```,```ape``` these files are all treated as ```WAVE```.

**Comment by lifegpc** [comment]
> I will check it more in depth later. Things can probably be simplified a bit.
> 
> Maybe done already. If a file has just one chapter, then pli.m_cue (and related values) should be set to false. To avoid unnecessary parsing of cue again later.

All done.
I think if it can not find cover with audio file, it can also try to find cover reletived to cue sheet file name. This will better, if there are multiply file in a folder but use same cover.

**Comment by lifegpc** [comment]
I add a new flag that control whether to parse cue sheet again. And also add an advanced settings.

**Comment by clsid2** [comment]
I would like to avoid storing much cue specific data in the playlist file.

You could add a generic "cover" value in the playlist file. When parsing the cue, you then only need to check for cover with same name once. Then for each file, check if it has a file specific cover, and if not use the global cover. This avoid the need to remember cue file and the option to reparse it. Plus people can use the cover entry for things unrelated to cue files as well.

**Comment by lifegpc** [comment]
> I would like to avoid storing much cue specific data in the playlist file.
> 
> You could add a generic "cover" value in the playlist file. When parsing the cue, you then only need to check for cover with same name once. Then for each file, check if it has a file specific cover, and if not use the global cover. This avoid the need to remember cue file and the option to reparse it. Plus people can use the cover entry for things unrelated to cue files as well.

done.

File: src/mpc-hc/PlayerPlaylistBar.cpp [hunk 2]
```diff
@@ -553,6 +716,13 @@ bool CPlayerPlaylistBar::ParseMPCPlayList(CString fn)
                 pli[i].m_ainput = _ttol(value);
             } else if (key == _T("country")) {
                 pli[i].m_country = _ttol(value);
+            } else if (key == _T("m_cue_filename")) {
```
Comment (clsid2) [inline src/mpc-hc/PlayerPlaylistBar.cpp]: Better to use a bit more pretty names:
"cuefile"
"cue_index"

**Comment by clsid2** [comment]
@adipose
Do you have any additional suggestions for further improvement?

**Comment by lifegpc** [comment]
done.

**Comment by adipose** [comment]
image

Is this meant to check the file extension?  If so it should use the last "." and compare to file extension. There are multiple examples in the code of this. Also, if so, "wave" is not the common extension, "wav" is, for wave audio (compressed and uncompressed).

**Comment by adipose** [comment]
Looks like "WAVE" is valid, though it should be parsed by space technically, and not by trailing characters.

**Comment by lifegpc** [comment]
There are only 5 valid file entry type in cue sheet: ```WAVE```, ```MP3```, ```AIFF```, ```BINARY``` and ```MOTOROLA```. I don't think it will have a BUG.

**Comment by adipose** [comment]
I understand.  It will probably work.  I just prefer to treat tokens as tokens rather than ends of strings.  For example, a cue sheet line ending with "BIGWAVE" would also match your algorithm despite not being a valid line.

Not a big concern.

File: src/mpc-hc/MainFrm.cpp [hunk 3]
```diff
@@ -12703,19 +12818,31 @@ void CMainFrame::OpenSetupWindowTitle(bool reset /*= false*/)
...
                                     title = CString(bstr.m_str);
+                                    hasName = true;
                                     break;
                                 }
                             }
                         }
                         EndEnumFilters;
                     }
+
+                    if (!hasName && pli->m_cue) {
```
Comment (clsid2) [inline src/mpc-hc/MainFrm.cpp]: I think it is better to remove the m_cue_metadata global variable and also this piece of code. This same title information is also set as the label of the playlist entry, right?

We should follow the option to use title (Options > Player > Title bar > replace filename with title). I might enable that by default in another patch.

So higher up this change could be made:
```
bool use_label = false;
CPlaylistItem* pli = m_wndPlaylistBar.GetCur();
if (pli && pli->m_label && !pli->m_label.IsEmpty()) {
    if (s.fTitleBarTextTitle || pli->m_fns.GetHead().Left(4) == _T("http")) {
        title = pli->m_label;
        use_label = true;
    }
}
```

Comment (lifegpc) [reply src/mpc-hc/MainFrm.cpp]: Seems a good idea.

**Comment by lifegpc** [comment]
Done.

File: src/mpc-hc/MainFrm.h [hunk 4]
```diff
@@ -143,6 +143,11 @@ struct SubtitleInput {
         : pSubStream(pSubStream), pSourceFilter(pSourceFilter) {};
 };
 
+struct CueMetadata {
```
Comment (clsid2) [inline src/mpc-hc/MainFrm.h]: This is no longer used. So can be removed.

**Comment by lifegpc** [comment]
I find a small bug, which related to file encoding. I try open gb2312(native code page on my system) file, and get garbage char.

**Pull request merged**